akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntitiesShardStore.scala

```diff
@@ -99,8 +99,11 @@ private[akka] final class EventSourcedRememberEntitiesShardStore(
       def persistEventsAndHandleComplete(evts: List[StateChange]): Unit = {
         persistAll(evts) { _ =>
           left -= 1
-          if (left == 0) {
+          val isLastEvent = left == 0
+          if (isLastEvent) {
             sender() ! RememberEntitiesShardStore.UpdateDone(started, stopped)
+          }
+          if (isLastEvent || isSnapshotNeeded) {
             state = state.copy(state.entities.union(started).diff(stopped))
             saveSnapshotWhenNeeded()
```

Comment:
I think we should still only save the snapshot after the last event, because otherwise the snapshot represents something partial.

I'd suggest another `var` that is updated for each event and then for the last event used as the condition for saveSnapshotWhenNeeded.

```
var saveSnap = false
...
... {
  saveSnap = saveSnap || isSnapshotNeeded
  if (isLastEvent && saveSnap)
    saveSnapshot()
}
```

Reply:
Good idea. I thought the snapshot should be saved on the exact index but it would not represent the correct sequence number anyway including future events of the batch.

I think it is more clear and logically correct now.

Reply:
I thought once more why I first suggested a snapshot on the exact sequence number divided by snapshotAfter:

https://github.com/akka/akka/blob/73d74dfc69b26cc8dea5a6800ec6ad6cd28168f9/akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntitiesShardStore.scala#L129

We will break this logic with the suggestion. What do you think @patriknw ?

Reply:
Can't we always subtract a full batch size from the calculated fromOffset, that way it doesn't matter where in the batch the offset happened?

Reply:
@johanandren Do you mean just keep one additional batch for this special circumstance? I would be fine with that.

Reply:
Yes, that is the idea I had.